File: standalone-coming-soon/subscribe.py
#!/usr/bin/env python3
import csv
import html
import json
import os
import re
import smtplib
import sys
import time
from contextlib import contextmanager
from datetime import datetime, timezone
from email.mime.multipart import MIMEMultipart
from email.mime.text import MIMEText
from email.utils import formataddr, parseaddr
from urllib.parse import parse_qs

try:
    import fcntl
except ImportError:  # pragma: no cover
    fcntl = None


EMAIL_RE = re.compile(r"^[^\s@]+@[^\s@]+\.[^\s@]+$")
MAX_BODY_BYTES = 4096
RATE_LIMIT_WINDOW_SECONDS = 60 * 60
RATE_LIMIT_MAX_ATTEMPTS = 6


SCRIPT_DIR = os.path.dirname(os.path.abspath(__file__))
DATA_DIR = os.environ.get("COMING_SOON_DATA_DIR", os.path.join(SCRIPT_DIR, "data"))
SUBSCRIBERS_FILE = os.environ.get(
    "COMING_SOON_SUBSCRIBERS_FILE",
    os.path.join(DATA_DIR, "subscribers.csv"),
)
RATE_LIMIT_FILE = os.path.join(DATA_DIR, "rate-limit.json")
LOCK_FILE = os.path.join(DATA_DIR, ".subscribe.lock")


def respond(status_code, payload):
    reason = {
        200: "OK",
        400: "Bad Request",
        405: "Method Not Allowed",
        413: "Payload Too Large",
        429: "Too Many Requests",
        500: "Internal Server Error",
        502: "Bad Gateway",
    }.get(status_code, "OK")
    print(f"Status: {status_code} {reason}")
    print("Content-Type: application/json; charset=utf-8")
    print("Cache-Control: no-store")
    print()
    print(json.dumps(payload))


def fail(status_code, message):
    respond(status_code, {"success": False, "error": message})


@contextmanager
def file_lock():
    os.makedirs(DATA_DIR, exist_ok=True)
    with open(LOCK_FILE, "a", encoding="utf-8") as lock:
        if fcntl:
            fcntl.flock(lock.fileno(), fcntl.LOCK_EX)
        try:
            yield
        finally:
            if fcntl:
                fcntl.flock(lock.fileno(), fcntl.LOCK_UN)


def read_form_value(form, name):
    value = form.get(name, "")
    return value.strip() if isinstance(value, str) else ""


def parse_form(content_length):
    content_type = os.environ.get("CONTENT_TYPE", "").split(";", 1)[0].strip().lower()
    if content_type != "application/x-www-form-urlencoded":
        raise ValueError("Unsupported form content type.")

    raw_body = sys.stdin.buffer.read(content_length).decode("utf-8", "replace")
    parsed = parse_qs(raw_body, keep_blank_values=True)
    return {
        key: values[0]
        for key, values in parsed.items()
        if values
    }


def get_client_ip():
    forwarded_for = os.environ.get("HTTP_X_FORWARDED_FOR", "")
    if forwarded_for:
        return forwarded_for.split(",")[0].strip()[:64]
    return os.environ.get("REMOTE_ADDR", "unknown")[:64]


def get_user_agent():
    return os.environ.get("HTTP_USER_AGENT", "unknown")[:300]


def normalize_email(value):
    return value.strip().lower()


def validate_email(value):
    return bool(value and len(value) <= 254 and EMAIL_RE.match(value))


def load_rate_limits():
    try:
        with open(RATE_LIMIT_FILE, "r", encoding="utf-8") as handle:
            data = json.load(handle)
            return data if isinstance(data, dict) else {}
    except FileNotFoundError:
        return {}
    except json.JSONDecodeError:
        return {}


def save_rate_limits(data):
    with open(RATE_LIMIT_FILE, "w", encoding="utf-8") as handle:
        json.dump(data, handle)


def check_rate_limit(ip_address):
    now = int(time.time())
    since = now - RATE_LIMIT_WINDOW_SECONDS
    data = load_rate_limits()
    attempts = [stamp for stamp in data.get(ip_address, []) if isinstance(stamp, int) and stamp >= since]

    if len(attempts) >= RATE_LIMIT_MAX_ATTEMPTS:
        data[ip_address] = attempts
        save_rate_limits(data)
        return False

    attempts.append(now)
    data[ip_address] = attempts

    for key in list(data.keys()):
        data[key] = [
            stamp for stamp in data.get(key, []) if isinstance(stamp, int) and stamp >= since
        ]
        if not data[key]:
            del data[key]

    save_rate_limits(data)
    return True


def load_existing_emails():
    try:
        with open(SUBSCRIBERS_FILE, "r", newline="", encoding="utf-8") as handle:
            return {
                row.get("email", "").strip().lower()
                for row in csv.DictReader(handle)
                if row.get("email")
            }
    except FileNotFoundError:
        return set()


def append_subscriber(email_address, ip_address, user_agent):
    file_exists = os.path.exists(SUBSCRIBERS_FILE)
    with open(SUBSCRIBERS_FILE, "a", newline="", encoding="utf-8") as handle:
        writer = csv.DictWriter(
            handle,
            fieldnames=["created_at_utc", "email", "ip_address", "user_agent"],
        )
        if not file_exists:
            writer.writeheader()
        writer.writerow(
            {
                "created_at_utc": datetime.now(timezone.utc).isoformat(),
                "email": email_address,
                "ip_address": ip_address,
                "user_agent": user_agent,
            }
        )


def smtp_config():
    host = os.environ.get("SMTP_HOST", "smtp-relay.brevo.com")
    port = int(os.environ.get("SMTP_PORT", "587"))
    username = os.environ.get("SMTP_USER")
    password = os.environ.get("SMTP_PASS")
    sender = os.environ.get("SMTP_FROM")
    admin_email = os.environ.get("ADMIN_EMAIL")

    missing = [
        name
        for name, value in {
            "SMTP_USER": username,
            "SMTP_PASS": password,
            "SMTP_FROM": sender,
            "ADMIN_EMAIL": admin_email,
        }.items()
        if not value
    ]
    if missing:
        raise RuntimeError("Missing email configuration: " + ", ".join(missing))

    return {
        "host": host,
        "port": port,
        "username": username,
        "password": password,
        "sender": sender,
        "admin_email": admin_email,
    }


def brand_html(content):
    return f"""\
<!doctype html>
<html>
  <body style="margin:0;background:#f4f7f5;padding:24px;font-family:Arial,Helvetica,sans-serif;color:#122018;">
    <table role="presentation" width="100%" cellspacing="0" cellpadding="0" style="border-collapse:collapse;">
      <tr>
        <td align="center">
          <table role="presentation" width="100%" cellspacing="0" cellpadding="0" style="max-width:560px;border-collapse:collapse;background:#ffffff;border:1px solid #dfe8e2;border-radius:12px;overflow:hidden;">
            <tr>
              <td style="padding:28px 30px 10px;">
                <p style="margin:0;color:#28b968;font-size:12px;font-weight:700;letter-spacing:2px;text-transform:uppercase;">NomiTips</p>
              </td>
            </tr>
            <tr>
              <td style="padding:0 30px 30px;">
                {content}
              </td>
            </tr>
          </table>
        </td>
      </tr>
    </table>
  </body>
</html>
"""


def build_subscriber_email(site_url):
    safe_url = html.escape(site_url, quote=True)
    html_body = brand_html(
        f"""
        <h1 style="margin:12px 0 14px;font-size:28px;line-height:1.15;color:#0b130f;">You're on the early access list.</h1>
        <p style="margin:0 0 16px;color:#48564d;font-size:16px;line-height:1.65;">
          Thank you for joining NomiTips before launch. We are preparing a focused fitness platform for women with structured programs, coaching support, progress tracking, and practical nutrition guidance.
        </p>
        <div style="margin:22px 0;padding:18px;border-radius:10px;background:#edfdf3;border:1px solid #c5f6d8;">
          <p style="margin:0;color:#17653d;font-size:15px;line-height:1.55;font-weight:700;">
            Early subscribers may receive a special launch bonus or exclusive benefit when the platform opens.
          </p>
        </div>
        <p style="margin:0 0 22px;color:#48564d;font-size:16px;line-height:1.65;">
          We will email you when early access begins.
        </p>
        <a href="{safe_url}" style="display:inline-block;background:#111c16;color:#ffffff;text-decoration:none;border-radius:8px;padding:12px 18px;font-weight:700;">
          Visit NomiTips
        </a>
        <p style="margin:26px 0 0;color:#7a877f;font-size:13px;line-height:1.5;">
          You are receiving this because this email address joined the NomiTips launch list.
        </p>
        """
    )
    text_body = (
        "You're on the NomiTips early access list.\n\n"
        "Thank you for joining before launch. NomiTips is preparing structured programs, "
        "coaching support, progress tracking, and practical nutrition guidance.\n\n"
        "Early subscribers may receive a special launch bonus or exclusive benefit when "
        "the platform opens.\n\n"
        f"Visit: {site_url}\n"
    )
    return text_body, html_body


def build_admin_email(email_address, ip_address, user_agent):
    timestamp = datetime.now(timezone.utc).strftime("%Y-%m-%d %H:%M:%S UTC")
    safe_email = html.escape(email_address)
    safe_ip = html.escape(ip_address)
    safe_agent = html.escape(user_agent)
    html_body = brand_html(
        f"""
        <h1 style="margin:12px 0 14px;font-size:26px;line-height:1.2;color:#0b130f;">New early access subscriber</h1>
        <p style="margin:0 0 18px;color:#48564d;font-size:16px;line-height:1.65;">
          A visitor joined the NomiTips Coming Soon early access list.
        </p>
        <table role="presentation" cellspacing="0" cellpadding="0" style="width:100%;border-collapse:collapse;background:#f6faf7;border:1px solid #e0e9e3;border-radius:10px;">
          <tr><td style="padding:14px 16px;color:#718078;width:130px;">Email</td><td style="padding:14px 16px;color:#122018;font-weight:700;">{safe_email}</td></tr>
          <tr><td style="padding:14px 16px;color:#718078;border-top:1px solid #e0e9e3;">Date</td><td style="padding:14px 16px;color:#122018;border-top:1px solid #e0e9e3;">{timestamp}</td></tr>
          <tr><td style="padding:14px 16px;color:#718078;border-top:1px solid #e0e9e3;">IP address</td><td style="padding:14px 16px;color:#122018;border-top:1px solid #e0e9e3;">{safe_ip}</td></tr>
          <tr><td style="padding:14px 16px;color:#718078;border-top:1px solid #e0e9e3;">User agent</td><td style="padding:14px 16px;color:#122018;border-top:1px solid #e0e9e3;">{safe_agent}</td></tr>
        </table>
        """
    )
    text_body = (
        "New NomiTips early access subscriber\n\n"
        f"Email: {email_address}\n"
        f"Date: {timestamp}\n"
        f"IP address: {ip_address}\n"
        f"User agent: {user_agent}\n"
    )
    return text_body, html_body


def send_email(config, to_address, subject, text_body, html_body):
    sender_name, sender_email = parseaddr(config["sender"])
    if not sender_email:
        sender_email = config["username"]
    formatted_sender = formataddr((sender_name or "NomiTips", sender_email))

    message = MIMEMultipart("alternative")
    message["From"] = formatted_sender
    message["To"] = to_address
    message["Subject"] = subject
    message.attach(MIMEText(text_body, "plain", "utf-8"))
    message.attach(MIMEText(html_body, "html", "utf-8"))

    smtp_client = smtplib.SMTP_SSL if config["port"] == 465 else smtplib.SMTP
    with smtp_client(config["host"], config["port"], timeout=20) as server:
        if config["port"] != 465:
            server.starttls()
        server.login(config["username"], config["password"])
        server.sendmail(sender_email, [to_address], message.as_string())


def handle_request():
    if os.environ.get("REQUEST_METHOD", "GET").upper() != "POST":
        fail(405, "This endpoint only accepts POST requests.")
        return

    try:
        content_length = int(os.environ.get("CONTENT_LENGTH") or "0")
    except ValueError:
        content_length = 0

    if content_length > MAX_BODY_BYTES:
        fail(413, "The submitted form is too large.")
        return

    try:
        form = parse_form(content_length)
    except ValueError:
        fail(400, "The submitted form could not be processed.")
        return

    if read_form_value(form, "company"):
        respond(200, {"success": True})
        return

    email_address = normalize_email(read_form_value(form, "email"))
    if not validate_email(email_address):
        fail(400, "Please enter a valid email address.")
        return

    ip_address = get_client_ip()
    user_agent = get_user_agent()
    site_url = os.environ.get("SITE_URL", "https://nomitips.com")

    try:
        with file_lock():
            if not check_rate_limit(ip_address):
                fail(429, "Too many attempts. Please try again later.")
                return

            if email_address in load_existing_emails():
                respond(200, {"success": True})
                return

            config = smtp_config()
            subscriber_text, subscriber_html = build_subscriber_email(site_url)
            admin_text, admin_html = build_admin_email(email_address, ip_address, user_agent)

            send_email(
                config,
                email_address,
                "You're on the NomiTips early access list",
                subscriber_text,
                subscriber_html,
            )
            send_email(
                config,
                config["admin_email"],
                "New Early Access Signup - NomiTips",
                admin_text,
                admin_html,
            )
            append_subscriber(email_address, ip_address, user_agent)

        respond(200, {"success": True})
    except RuntimeError as error:
        print(f"Configuration error: {error}", file=sys.stderr)
        fail(500, "Email is not configured yet. Please try again later.")
    except Exception as error:
        print(f"Subscription error: {error}", file=sys.stderr)
        fail(502, "We could not complete your subscription. Please try again.")


if __name__ == "__main__":
    handle_request()

File: standalone-coming-soon/subscribe.php
<?php
/**
 * NomiTips Coming Soon - Subscription Handler
 * Deploy this to Namecheap cPanel hosting
 */

// ─── Configuration ──────────────────────────────────────────
// Set these in your cPanel or hardcode for testing
$SMTP_HOST    = getenv('SMTP_HOST')    ?: 'smtp-relay.brevo.com';
$SMTP_PORT    = getenv('SMTP_PORT')    ?: 587;
$SMTP_USER    = getenv('SMTP_USER')    ?: 'b6b4e5001@smtp-brevo.com';
$SMTP_PASS    = getenv('SMTP_PASS')    ?: 'xsmtpsib-fd8d2aea925067bca7273c6f7bc5c23ccfa37fc120d5639f7052b5c51ce7c025-Ezy2NSAqwX1rLdZ8';
$SMTP_FROM    = getenv('SMTP_FROM')    ?: 'NomiTips <batsindakeynesbenoit10101@gmail.com>';
$ADMIN_EMAIL  = getenv('ADMIN_EMAIL')  ?: 'batsindakeynesbenoit10101@gmail.com';
$SITE_URL     = getenv('SITE_URL')     ?: 'https://nomitips.com';
$DATA_DIR     = __DIR__ . '/data';
$SUBSCRIBERS  = $DATA_DIR . '/subscribers.csv';
$LOGO_URL     = 'https://raw.githubusercontent.com/NomiTips/assets/main/logo-email.png';

// ─── Helper Functions ───────────────────────────────────────
function respond($code, $data) {
    http_response_code($code);
    header('Content-Type: application/json; charset=utf-8');
    header('Cache-Control: no-store');
    echo json_encode($data);
    exit;
}

function validate_email($email) {
    return filter_var($email, FILTER_VALIDATE_EMAIL) && strlen($email) <= 254;
}

function is_bot($company) {
    return !empty($company);
}

function save_subscriber($email) {
    global $DATA_DIR, $SUBSCRIBERS;
    if (!is_dir($DATA_DIR)) mkdir($DATA_DIR, 0755, true);
    $file = fopen($SUBSCRIBERS, 'a');
    if ($file) {
        fputcsv($file, ['created_at_utc', 'email']);
        fputcsv($file, [date('c'), $email]);
        fclose($file);
    }
}

// ─── SMTP via PHP Sockets ──────────────────────────────────
function smtp_send($to, $subject, $textBody, $htmlBody) {
    global $SMTP_HOST, $SMTP_PORT, $SMTP_USER, $SMTP_PASS, $SMTP_FROM;

    list($fromName, $fromEmail) = explode(' <', str_replace('>', '', $SMTP_FROM));

    $headers  = "From: =?UTF-8?B?" . base64_encode($fromName) . "?= <$fromEmail>\r\n";
    $headers .= "To: $to\r\n";
    $headers .= "Subject: =?UTF-8?B?" . base64_encode($subject) . "?=\r\n";
    $headers .= "MIME-Version: 1.0\r\n";
    $headers .= "Content-Type: multipart/alternative; boundary=\"boundary\"\r\n";
    $headers .= "\r\n";

    $body = "--boundary\r\n";
    $body .= "Content-Type: text/plain; charset=UTF-8\r\n\r\n";
    $body .= $textBody . "\r\n\r\n";
    $body .= "--boundary\r\n";
    $body .= "Content-Type: text/html; charset=UTF-8\r\n\r\n";
    $body .= $htmlBody . "\r\n\r\n";
    $body .= "--boundary--";

    $errno = 0;
    $errstr = '';
    $fp = fsockopen("ssl://$SMTP_HOST", $SMTP_PORT, $errno, $errstr, 30);
    if (!$fp) {
        $fp = fsockopen($SMTP_HOST, $SMTP_PORT, $errno, $errstr, 30);
    }
    if (!$fp) throw new Exception("SMTP connection failed: $errstr");

    $response = fgets($fp, 512);

    // EHLO
    fwrite($fp, "EHLO nomitips.com\r\n");
    while ($response = fgets($fp, 512)) {
        if (strpos($response, '250') === 0 && substr($response, 3, 1) === ' ') break;
    }

    // STARTTLS
    fwrite($fp, "STARTTLS\r\n");
    $response = fgets($fp, 512);
    if (strpos($response, '220') === 0) {
        stream_context_set_option($fp, 'ssl', 'verify_peer', false);
        stream_context_set_option($fp, 'ssl', 'verify_peer_name', false);
        $crypto = stream_socket_enable_crypto($fp, true, STREAM_CRYPTO_METHOD_TLSv1_2_CLIENT);
        if (!$crypto) throw new Exception("TLS failed");

        fwrite($fp, "EHLO nomitips.com\r\n");
        while ($response = fgets($fp, 512)) {
            if (strpos($response, '250') === 0 && substr($response, 3, 1) === ' ') break;
        }
    }

    // AUTH LOGIN
    fwrite($fp, "AUTH LOGIN\r\n");
    fgets($fp, 512);
    fwrite($fp, base64_encode($SMTP_USER) . "\r\n");
    fgets($fp, 512);
    fwrite($fp, base64_encode($SMTP_PASS) . "\r\n");
    $response = fgets($fp, 512);
    if (strpos($response, '235') !== 0) throw new Exception("SMTP auth failed: $response");

    // MAIL FROM
    fwrite($fp, "MAIL FROM:<$fromEmail>\r\n");
    fgets($fp, 512);

    // RCPT TO
    fwrite($fp, "RCPT TO:<$to>\r\n");
    fgets($fp, 512);

    // DATA
    fwrite($fp, "DATA\r\n");
    fgets($fp, 512);
    fwrite($fp, $headers . $body . "\r\n.\r\n");
    fgets($fp, 512);

    // QUIT
    fwrite($fp, "QUIT\r\n");
    fclose($fp);

    return true;
}

// ─── Email Templates ────────────────────────────────────────
function subscriber_email($email) {
    global $SITE_URL, $LOGO_URL;
    $url = htmlspecialchars($SITE_URL, ENT_QUOTES);
    $html = <<<HTML
<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8"><meta name="viewport" content="width=device-width, initial-scale=1.0"></head>
<body style="margin:0;padding:0;background-color:#0a0f0d;font-family:'Helvetica Neue',Helvetica,Arial,sans-serif;">
  <table role="presentation" width="100%" cellspacing="0" cellpadding="0" border="0" style="background-color:#0a0f0d;">
    <tr><td align="center" style="padding:40px 20px;">
      <table role="presentation" width="100%" cellspacing="0" cellpadding="0" border="0" style="max-width:520px;background-color:#111a14;border-radius:16px;border:1px solid #1e2e24;overflow:hidden;">

        <!-- Logo -->
        <tr><td align="center" style="padding:40px 30px 20px;">
          <img src="$LOGO_URL" alt="NomiTips" width="200" style="max-width:200px;height:auto;display:block;margin:0 auto;">
        </td></tr>

        <tr><td style="padding:0 40px;"><div style="height:1px;background:linear-gradient(90deg,transparent,#28b968,transparent);"></div></td></tr>

        <!-- Content -->
        <tr><td style="padding:30px 40px 10px;">
          <h1 style="margin:0 0 20px;font-size:26px;font-weight:700;color:#ffffff;line-height:1.3;">Welcome to the Family 🎉</h1>
          <p style="margin:0 0 16px;font-size:16px;color:#a8c4b8;line-height:1.7;">
            You're officially on the NomiTips early access list. We're building something special — a premium fitness experience designed specifically for women who want real, lasting results.
          </p>
        </td></tr>

        <!-- What's Coming -->
        <tr><td style="padding:10px 40px;">
          <div style="background-color:#0d1610;border:1px solid #1e2e24;border-radius:12px;padding:24px;">
            <p style="margin:0 0 14px;font-size:14px;color:#28b968;font-weight:700;letter-spacing:1px;text-transform:uppercase;">What's Coming</p>
            <table role="presentation" width="100%" cellspacing="0" cellpadding="0" border="0">
              <tr><td style="padding:6px 0;color:#a8c4b8;font-size:15px;">✦&nbsp; Custom workout programs</td></tr>
              <tr><td style="padding:6px 0;color:#a8c4b8;font-size:15px;">✦&nbsp; 1-on-1 coaching support</td></tr>
              <tr><td style="padding:6px 0;color:#a8c4b8;font-size:15px;">✦&nbsp; Nutrition planning tools</td></tr>
              <tr><td style="padding:6px 0;color:#a8c4b8;font-size:15px;">✦&nbsp; Progress tracking & analytics</td></tr>
              <tr><td style="padding:6px 0;color:#a8c4b8;font-size:15px;">✦&nbsp; Community & transformation support</td></tr>
            </table>
          </div>
        </td></tr>

        <!-- Bonus Card -->
        <tr><td style="padding:20px 40px 10px;">
          <div style="background:linear-gradient(135deg,#0d1610 0%,#142218 100%);border:1px solid #28b968;border-radius:12px;padding:24px;text-align:center;">
            <p style="margin:0 0 8px;font-size:24px;">🎁</p>
            <p style="margin:0 0 8px;font-size:16px;color:#ffffff;font-weight:700;">Early Subscriber Exclusive</p>
            <p style="margin:0;font-size:14px;color:#28b968;font-weight:600;">You'll receive a special launch reward when we go live</p>
          </div>
        </td></tr>

        <!-- CTA -->
        <tr><td align="center" style="padding:30px 40px;">
          <a href="$url" style="display:inline-block;background-color:#28b968;color:#0a0f0d;text-decoration:none;border-radius:10px;padding:14px 32px;font-size:15px;font-weight:700;">Visit NomiTips</a>
        </td></tr>

        <tr><td style="padding:0 40px 10px;"><div style="height:1px;background-color:#1e2e24;"></div></td></tr>
        <tr><td style="padding:10px 40px 30px;">
          <p style="margin:0 0 8px;font-size:12px;color:#4a6356;text-align:center;">You're receiving this because you joined the NomiTips launch list.</p>
          <p style="margin:0;font-size:12px;color:#4a6356;text-align:center;">NomiTips &copy; 2026 &middot; All rights reserved</p>
        </td></tr>

      </table>
    </td></tr>
  </table>
</body>
</html>
HTML;
    $text = "Welcome to NomiTips Early Access!\n\nYou're officially on the early access list.\n\nWHAT'S COMING:\n• Custom workout programs\n• 1-on-1 coaching support\n• Nutrition planning tools\n• Progress tracking & analytics\n\nVisit: $url\n";
    return ['text' => $text, 'html' => $html];
}

function admin_email($email) {
    global $LOGO_URL;
    $safe = htmlspecialchars($email, ENT_QUOTES);
    $date = date('Y-m-d H:i:s');
    $html = <<<HTML
<!DOCTYPE html>
<html lang="en">
<head><meta charset="utf-8"></head>
<body style="margin:0;padding:0;background-color:#0a0f0d;font-family:'Helvetica Neue',Helvetica,Arial,sans-serif;">
  <table role="presentation" width="100%" cellspacing="0" cellpadding="0" border="0" style="background-color:#0a0f0d;">
    <tr><td align="center" style="padding:40px 20px;">
      <table role="presentation" width="100%" cellspacing="0" cellpadding="0" border="0" style="max-width:520px;background-color:#111a14;border-radius:16px;border:1px solid #1e2e24;overflow:hidden;">
        <tr><td style="padding:30px 40px;">
          <p style="margin:0 0 4px;font-size:11px;color:#4a6356;letter-spacing:2px;text-transform:uppercase;">NomiTips Admin</p>
          <h1 style="margin:0 0 20px;font-size:22px;color:#ffffff;">New Early Access Signup</h1>
          <table role="presentation" width="100%" cellspacing="0" cellpadding="0" border="0" style="background-color:#0d1610;border:1px solid #1e2e24;border-radius:10px;">
            <tr><td style="padding:14px 16px;color:#4a6356;width:100px;font-size:14px;">Email</td><td style="padding:14px 16px;color:#ffffff;font-weight:600;font-size:14px;">$safe</td></tr>
            <tr><td style="padding:14px 16px;color:#4a6356;border-top:1px solid #1e2e24;font-size:14px;">Date</td><td style="padding:14px 16px;color:#ffffff;border-top:1px solid #1e2e24;font-size:14px;">$date</td></tr>
          </table>
        </td></tr>
      </table>
    </td></tr>
  </table>
</body>
</html>
HTML;
    $text = "New NomiTips early access subscriber\n\nEmail: $email\nDate: $date\n";
    return ['text' => $text, 'html' => $html];
}

// ─── Handle Request ─────────────────────────────────────────
if ($_SERVER['REQUEST_METHOD'] !== 'POST') {
    respond(405, ['success' => false, 'error' => 'POST only.']);
}

$company = $_POST['company'] ?? '';
if (is_bot($company)) {
    respond(200, ['success' => true]);
}

$email = strtolower(trim($_POST['email'] ?? ''));
if (!validate_email($email)) {
    respond(400, ['success' => false, 'error' => 'Please enter a valid email address.']);
}

try {
    save_subscriber($email);

    $sub = subscriber_email($email);
    smtp_send($email, "You're on the NomiTips Early Access List 🎉", $sub['text'], $sub['html']);

    $adm = admin_email($email);
    smtp_send($ADMIN_EMAIL, "New Early Access Signup — NomiTips", $adm['text'], $adm['html']);

    respond(200, ['success' => true]);
} catch (Exception $e) {
    error_log("NomiTips subscribe error: " . $e->getMessage());
    respond(500, ['success' => false, 'error' => 'Something went wrong. Please try again.']);
}
